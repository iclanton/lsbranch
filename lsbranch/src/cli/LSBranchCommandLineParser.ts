// Copyright (c) Ian Clanton-Thuon. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { CommandLineParser, CommandLineStringParameter } from '@rushstack/ts-command-line';
import { Terminal, ConsoleTerminalProvider, AlreadyReportedError } from '@rushstack/node-core-library';

import { ILSBranchActionBaseOptions } from './actions/LSBranchActionBase';
import { LSAction } from './actions/LSAction';
import { AddAction } from './actions/AddAction';
import { LSBranchConfig } from '../config/LSBranchConfig';

export const LSBRANCH_TOOL_FILENAME: string = 'lsbranch';

export class LSBranchCommandLineParser extends CommandLineParser {
  private _terminal: Terminal;
  private _configPathParameter!: CommandLineStringParameter;

  public get terminal(): Terminal {
    return this._terminal;
  }

  public constructor() {
    super({
      toolFilename: LSBRANCH_TOOL_FILENAME,
      toolDescription: 'LSBranch is a simple tool for listing the active branches in git clones.'
    });

    this._terminal = new Terminal(new ConsoleTerminalProvider());

    const actionOptions: ILSBranchActionBaseOptions = {
      terminal: this.terminal,
      config: new LSBranchConfig(() => this._configPathParameter.value)
    };

    this.addAction(new LSAction(actionOptions));
    this.addAction(new AddAction(actionOptions));
  }

  protected onDefineParameters(): void {
    this._configPathParameter = this.defineStringParameter({
      parameterLongName: '--config',
      parameterShortName: '-c',
      argumentName: 'PATH',
      description: 'Override the config file path. Defaults to ~/.lsbranchrc.json'
    });
  }

  protected async onExecute(): Promise<void> {
    process.exitCode = 1;
    try {
      await super.onExecute();
      process.exitCode = 0;
    } catch (e) {
      if (!(e instanceof AlreadyReportedError)) {
        this.terminal.writeErrorLine(`An unexpected error occurred: ${e}`);
      }
    }
  }
}
