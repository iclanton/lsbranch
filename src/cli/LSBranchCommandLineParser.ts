// Copyright (c) Ian Clanton-Thuon. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { CommandLineParser, CommandLineStringParameter } from '@rushstack/ts-command-line';
import { Terminal, ConsoleTerminalProvider, AlreadyReportedError } from '@rushstack/node-core-library';

import { LSAction } from './actions/LSAction';

export class LSBranchCommandLineParser extends CommandLineParser {
  private _terminal: Terminal;
  private _configPathParameter!: CommandLineStringParameter;

  public get terminal(): Terminal {
    return this._terminal;
  }

  public constructor() {
    super({
      toolFilename: 'lsbranch',
      toolDescription: 'LSBranch is a simple tool for listing the active branches in git clones.'
    });

    this._terminal = new Terminal(new ConsoleTerminalProvider());

    this.addAction(new LSAction({ terminal: this.terminal, configPath: this._configPathParameter.value }));
  }

  protected onDefineParameters(): void {
    this._configPathParameter = this.defineStringParameter({
      parameterLongName: '--config',
      parameterShortName: '-c',
      argumentName: 'PATH',
      description: 'Override the config file path. Defaults to ~/.lsbranch.json'
    });
  }

  protected async onExecute(): Promise<void> {
    process.exitCode = 1;
    try {
      await super.onExecute();
      process.exitCode = 0;
    } catch (e) {
      if (!(e instanceof AlreadyReportedError)) {
        this.terminal.writeErrorLine(`An unexpected error occurred: ${e}`);
      }
    }
  }
}
