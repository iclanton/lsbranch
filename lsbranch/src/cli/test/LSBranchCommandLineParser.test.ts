// Copyright (c) Ian Clanton-Thuon. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { AnsiEscape } from '@rushstack/node-core-library';
import { LSBranchCommandLineParser } from '../LSBranchCommandLineParser';

describe(LSBranchCommandLineParser.name, () => {
  let parser: LSBranchCommandLineParser;

  beforeEach(() => {
    // ts-command-line calls process.exit() which interferes with Jest
    jest.spyOn(process, 'exit').mockImplementation((code) => {
      throw new Error(`Test code called process.exit(${code})`);
    });

    // This call may terminate the entire test run because it invokes process.exit()
    // if it encounters errors.
    // TODO Remove the calls to process.exit() or override them for testing.
    parser = new LSBranchCommandLineParser();
  });

  it('prints the global help', () => {
    const helpText: string = AnsiEscape.formatForTests(parser.renderHelpText());
    expect(helpText).toMatchSnapshot();
  });

  it('prints the help for each action', () => {
    for (const action of parser.actions) {
      const helpText: string = AnsiEscape.formatForTests(action.renderHelpText());
      expect(helpText).toMatchSnapshot(action.actionName);
    }
  });
});
