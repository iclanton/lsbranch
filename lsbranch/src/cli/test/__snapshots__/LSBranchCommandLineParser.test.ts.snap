// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LSBranchCommandLineParser prints the global help 1`] = `
"usage: lsbranch [-h] [-c PATH] <command> ...

LSBranch is a simple tool for listing the active branches in git clones.

Positional arguments:
  <command>
    ls                  List branches.
    add                 Add a repo.

Optional arguments:
  -h, --help            Show this help message and exit.
  -c PATH, --config PATH
                        Override the config file path. Defaults to ~/.
                        lsbranchrc.json

[bold]For detailed help about a specific command, use: lsbranch <command> 
-h[normal]
"
`;

exports[`LSBranchCommandLineParser prints the help for each action: add 1`] = `
"usage: lsbranch add [-h] --path REPO_PATH [--alias ALIAS] [--no-validate]

Optional arguments:
  -h, --help        Show this help message and exit.
  --path REPO_PATH  The path to the repo root
  --alias ALIAS     The repo's alias
  --no-validate     If specified, do not ensure that the repo being added 
                    exists and is valid.
"
`;

exports[`LSBranchCommandLineParser prints the help for each action: ls 1`] = `
"usage: lsbranch ls [-h] [--json] [-a]

Optional arguments:
  -h, --help  Show this help message and exit.
  --json      If specified, present data in JSON format.
  -a, --all   If specified, display all local branches.
"
`;
